### PR TITLE
Importance score evaluation with a ranking correlation function

### PR DESCRIPTION
**Pull request by sylvainma**
Introduce a more robust way to evaluate video summarization models, following recommendations from the paper: [Rethinking the Evaluation of Video Summaries](https://arxiv.org/abs/1903.11328)

- [x] Add importance score evaluation methods using a ranking correlation metric
- [x] Link the score annotations in .h5 files to these new importance score evaluation methods
- [x] Dataset agnostic aggregation function (no longer `max` for SumMe and `avg` for TVSum)
- [x] Replace the baseline model by a random importance score model

### CODE REVIEW

**Comment by sylvainma** [comment]
After careful analysis, here is how we want to evaluate scores using the correlation mentioned in the paper:
* TVSum: we will use the original score annotation
* SumMe: despite the lack of score annotation, we will use the `/gtscore`

File: summarizer/datasets/correlation.ipynb [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+    }
+   ],
+   "source": [
+    "dataset = h5py.File(\"summarizer_dataset_summe_google_pool5.h5\", \"r\")\n",
+    "video_1 = dataset[\"video_1\"]\n",
+    "video_name = video_1[\"video_name\"][...].item()\n",
+    "n_frames = video_1[\"n_frames\"][...]\n",
+    "positions = video_1[\"picks\"][...]\n",
+    "\n",
+    "user_summary = video_1[\"user_summary\"][...] # (n_users, n_frames)\n",
```
Comment (msaintja) [inline summarizer/datasets/correlation.ipynb:214]: We should re-add the line `n_users, n_frames = user_summary.shape` after this, or redefine `n_users` in some way, otherwise the second cell following this one uses the `n_users` value from the TVSum dataset. This causes an out-of-bounds error (there are more in TVSum than SumMe).

File: summarizer/datasets/correlation.ipynb [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "_Observations:_\n",
+    "* Using `user_summary` is consistent: random scores leads to 0 and human is high (~0.22)\n",
+    "* Using `user_scores` matches the paper: random also near 0 and human high (us ~0.33)\n",
+    "\n",
+    "In the paper they did not experiment this with SumMe since there is no importance score annotations.\n",
+    "\n",
+    "Here, we should use `user_scores` (in other words `/gtscore`) to be consistent with TVSum, even if `/gtscore` is computed from summaries of annotators instead of absolute frame importance score. But, these summaries were made by annotators directly, without usiing KTS. So it is not biased by KTS. Which is the dark side of using F-score. So we can assume were're fine?"
```
Comment (msaintja) [inline summarizer/datasets/correlation.ipynb:339]:
```suggestion
    "Here, we should use `user_scores` (in other words `/gtscore`) to be consistent with TVSum, even if `/gtscore` is computed from summaries of annotators instead of absolute frame importance score. But, these summaries were made by annotators directly, without usiing KTS. So it is not biased by KTS. Which is the dark side of using F-score. So we can assume we're fine?"
```

File: summarizer/main.py [hunk 3]
```diff
@@ -12,36 +12,41 @@ def train(hps):
...
-        fscores_cv = []
+        corrs_cv, fscores_cv = [], []
         
         # Destination for weights and predictions on dataset
         weights_path = hps.weights_path[splits_file]
         pred_path = hps.pred_path[splits_file]
 
         # For every fold in current split file
-        fscore_max = 0.0
+        corr_max = 0.0
```
Comment (msaintja) [inline summarizer/main.py:22]:
```suggestion
        corr_max = -1.0
```

File: summarizer/models/dsn.py [hunk 4]
```diff
@@ -69,7 +69,7 @@ def train(self, fold):
         reward_writers = {key: [] for key in train_keys}
 
         # To record performances of the best epoch
-        best_f_score = 0.0
+        best_corr, best_f_score = 0.0, 0.0
```
Comment (msaintja) [inline summarizer/models/dsn.py:72]:
```suggestion
        best_corr, best_f_score = -1.0, 0.0
```

File: summarizer/models/logistic.py [hunk 5]
```diff
@@ -44,11 +43,10 @@ def train(self, fold):
         self.optimizer = torch.optim.Adam(
             self.model.parameters(),
             lr=self.hps.lr,
-            weight_decay=self.hps.l2_req
-        )
+            weight_decay=self.hps.l2_req)
 
         # To record performances of the best epoch
-        best_f_score = 0.0
+        best_corr, best_f_score = 0.0, 0.0
```
Comment (msaintja) [inline summarizer/models/logistic.py:49]:
```suggestion
        best_corr, best_f_score = -1.0, 0.0
```

File: summarizer/models/random.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    def train(self, fold):
+        self.model.train()
+        train_keys, _ = self._get_train_test_keys(fold)
+
+        criterion = nn.MSELoss()
+        if self.hps.use_cuda:
+            criterion = criterion.cuda()
+        
+        # To record performances of the best epoch
+        best_corr, best_f_score = 0.0, 0.0
```
Comment (msaintja) [inline summarizer/models/random.py:40]:
```suggestion
        best_corr, best_f_score = -1.0, 0.0
```

File: summarizer/models/sumgan.py [hunk 7]
```diff
@@ -327,7 +327,7 @@ def train(self, fold):
             self.loss_BCE.cuda()
 
         # To record performances of the best epoch
-        best_f_score = 0.0
+        best_corr, best_f_score = 0.0, 0.0
```
Comment (msaintja) [inline summarizer/models/sumgan.py:330]:
```suggestion
        best_corr, best_f_score = -1.0, 0.0
```

File: summarizer/models/transformer.py [hunk 8]
```diff
@@ -128,7 +128,7 @@ def train(self, fold):
         self.optimizer = torch.optim.Adam(parameters, lr=self.hps.lr, weight_decay=self.hps.l2_req)
 
         # To record performances of the best epoch
-        best_f_score = 0.0
+        best_corr, best_f_score = 0.0, 0.0
```
Comment (msaintja) [inline summarizer/models/transformer.py:131]:
```suggestion
        best_corr, best_f_score = -1.0, 0.0
```

File: summarizer/models/vasnet.py [hunk 9]
```diff
@@ -173,7 +173,7 @@ def train(self, fold):
         self.optimizer = torch.optim.Adam(parameters, lr=self.hps.lr, weight_decay=self.hps.l2_req)
 
         # To record performances of the best epoch
-        best_f_score = 0.0
+        best_corr, best_f_score = 0.0, 0.0
```
Comment (msaintja) [inline summarizer/models/vasnet.py:176]:
```suggestion
        best_corr, best_f_score = -1.0, 0.0
```

File: summarizer/datasets/correlation.ipynb (hunk 1 above)
Comment (sylvainma) [reply summarizer/datasets/correlation.ipynb:214]: Indeed thanks, fixed.
